File: Arrays/spiralTraverse.js
//? https://www.algoexpert.io/questions/Spiral%20Traverse
//! O(n) time | O(n) space
function spiralTraverse(array) {
	const result = [];
	let startRow = 0,
		endRow = array.length - 1;
	let startCol = 0, 
		endCol = array[0].length - 1;
	
		while(startRow <= endRow && startCol <= endCol) {
			for(let col = startCol; col <= endCol; col++) {
				result.push(array[startRow][col]);
			}
			for(let row = startRow + 1; row <= endRow; row++) {
				result.push(array[row][endCol]);
			}
			for(let col = endCol - 1; col >= startCol; col--) {
				/**
				 * Handle the edge case when there's a single row
				 * in the middle of the matrix. In this case, we don't
				 * want to double-count the values in this row, which 
				 * we've already counted in the first for loop above.
				 */
				if(startRow == endRow) break;
				result.push(array[endRow][col]);
			}
			for(let row = endRow - 1; row > startRow; row--) {
				/**
				 * Handle the edge case when there's a single column
				 * in the middle of the matrix. In this case, we don't
				 * want to double-count the values in this column, which
				 * we've already counted in the second for loop above.
				 */
				if(startCol == endCol) break;
				result.push(array[row][startCol]);
			} 
			startRow++;
			endRow--;
			startCol++;
			endCol--;

		} return result;
}

const array = [
//!  					   		[1, 2, 3, 4],
//! Test case for line no. 18	[10, 11, 12, 5],
//! startRow === endRow	   		[9, 8, 7, 6]

//!       						  [1, 2, 3],
//!       						  [12, 13, 4],
//!  Test case for line no. 23    11, 14, 5],
//!  startCol === endCol     	  [10, 15, 6],
//!       						  [9, 8, 7]

]

console.log(spiralTraverse(array));

File: Javascript-DSA/Arrays/spiralTraverse.cpp
#include <bits/stdc++.h>
using namespace std;

            //! using loops O(n) time | O(n) space
vector<int> spiralTraverse(vector<vector<int>> array) {

        
        vector <int> result = {};
        int startRow = 0, endRow = array.size() - 1;
        int startCol = 0, endCol = array[0].size() - 1;

        while ((startRow <= endRow) && (startCol <= endCol)) { 

                    for (int col = startCol; col <= endCol; col++) {
                        result.push_back(array[startRow][col]);
                    }
                     for (int row = startRow + 1; row <= endRow; row++) {
                         result.push_back(array[row][endCol]);
                     }
                     for (int col = endCol - 1; col >= startCol; col--) {
                        result.push_back(array[endRow][col]);
                     }
                     for(int row = endRow + 1; row < startRow; row++) {
                         result.push_back(array[row][startCol]);
                     }
                     startRow++;
                     endRow--;
                     startCol++;
                     endCol--
        }

        return result;


}

//! Using Recursion


int spiralTraverse(int array[][]) {
    vector<int> result = {};
    spiralFill(array, 0, array.size() - 1, 0, array[0].size() - 1, result);
    return result;

}

spiralFill(int array[][], int startRow, int endRow, int startCol, endCol, vector<int> result = {}) {

            if (startRow > endRow || startCol > endCol) {
                return;
            }

            for (int col = startCol; col <= endCol; col++)
            {
                result.push_back(array[startRow][col]);
            }
            for (int row = startRow + 1; row <= endRow; row++)
            {
                result.push_back(array[row][endCol]);
            }
            for (int col = endCol - 1; col >= startCol; col--)
            {
                result.push_back(array[endRow][col]);
            }
            for (int row = endRow + 1; row < startRow; row++)
            {
                result.push_back(array[row][startCol]);
            }
            spiralFill(int array[][], int startRow + 1, int endRow -1, int startCol + 1, endCol - 1, vector<int> result = {})
}   


